src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt

```diff
@@ -0,0 +1,147 @@
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import chip.devicecontroller.ChipClusters
+import chip.devicecontroller.ChipClusters.BasicCluster
+import chip.devicecontroller.ChipDeviceController
+import com.google.chip.chiptool.ChipClient
+import com.google.chip.chiptool.GenericChipDeviceListener
+import com.google.chip.chiptool.R
+import com.google.chip.chiptool.util.DeviceIdUtil
+import kotlinx.android.synthetic.main.basic_client_fragment.*
+import kotlinx.android.synthetic.main.basic_client_fragment.view.*
+import kotlinx.android.synthetic.main.on_off_client_fragment.*
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+class BasicClientFragment : Fragment() {
+  private val deviceController: ChipDeviceController
+    get() = ChipClient.getDeviceController(requireContext())
+
+  private val scope = CoroutineScope(Dispatchers.Main + Job())
+
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View {
+    return inflater.inflate(R.layout.basic_client_fragment, container, false).apply {
+      deviceController.setCompletionListener(ChipControllerCallback())
+
+      basicClusterUpdateAddressBtn.setOnClickListener { scope.launch { updateAddressClick() }}
+      readUserLabelBtn.setOnClickListener { scope.launch { sendReadUserLabelCommandClick() }}
+      writeUserLabelBtn.setOnClickListener { scope.launch { sendWriteUserLabelCommandClick() }}
+      readProductNameBtn.setOnClickListener { scope.launch { sendReadProductNameCommandClick() }}
+    }
+  }
+
+  override fun onStart() {
+    super.onStart()
+    // TODO: use the fabric ID that was used to commission the device
+    val testFabricId = "5544332211"
+    basicClusterFabricIdEd.setText(testFabricId)
+    basicClusterDeviceIdEd.setText(DeviceIdUtil.getLastDeviceId(requireContext()).toString())
+  }
+
+  inner class ChipControllerCallback : GenericChipDeviceListener() {
+    override fun onConnectDeviceComplete() {}
+
+    override fun onCommissioningComplete(nodeId: Long, errorCode: Int) {
+      Log.d(TAG, "onCommissioningComplete for nodeId $nodeId: $errorCode")
+    }
+
+    override fun onNotifyChipConnectionClosed() {
+      Log.d(TAG, "onNotifyChipConnectionClosed")
+    }
+
+    override fun onCloseBleComplete() {
+      Log.d(TAG, "onCloseBleComplete")
+    }
+
+    override fun onError(error: Throwable?) {
+      Log.d(TAG, "onError: $error")
+    }
+  }
+
+  override fun onStop() {
+    super.onStop()
+    scope.cancel()
+  }
+
+  private fun updateAddressClick() {
+    try {
+      deviceController.updateDevice(
+              basicClusterFabricIdEd.text.toString().toULong().toLong(),
+              basicClusterDeviceIdEd.text.toString().toULong().toLong()
```

Comment:
I think correct indentation here is two spaces past the beginning of `deviceController`?

Reply:
Thank you for feedback. fixed to two spaces